internal/controller/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +57,568 @@ type VolumeGroupReplicationReconciler struct {
 	Scheme *runtime.Scheme
 }
 
-//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplications,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplications,verbs=get;list;watch;update;patch
 //+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplications/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplications/finalizers,verbs=update
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplicationclasses,verbs=get;list;watch
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumegroupreplicationcontents,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumereplications,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups=replication.storage.openshift.io,resources=volumereplications/status,verbs=get;list
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims;persistentvolumes,verbs=get;list;watch;update
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims/finalizers,verbs=update
+
+/*
+Steps performed by the reconcile loop:
+- Fetch and validate the VGRClass CR
+- Fetch the matching PVCs based on the selector provided in the VGR CR, and check if they are already bounded to a CSI volume and the driver matches the driver provided in the VGRClass CR.
+- Annotate to the PVCs with owner and add the VGR finalizer to them.
+- Add the label selector to the VGR annotation, such that the PVC triggering a reconcile can fetch the VGR to reconcile
+- Create the VGRContent with the PVs list fetched above, add VGR name/namespace as the annotation to it
+- Wait for the volumes to be grouped, and the VGRContent to be updated with the group handle
+- Then, create the VR CR and add VGR name/namespace as the annotation to it
+- Update the VGR status with the VR status.
+
+In case of deletion:
```

Comment:
Handle deletion case of object first.
maybe define an instance object like https://github.com/csi-addons/kubernetes-csi-addons/blob/main/internal/controller/replication.storage/volumereplication_controller.go#L616
and then write  helper functions with it ? 

The reconcile function  is >250 lines.

Reply:
> Handle deletion case of object first.

I feel, if we handle deletion first then it becomes a bit arbritary while reading the code to understand what exactly is happening and why. Having it in the end tells the reader that these all operations happened when the CR was created and that's why all these cleanups are needed.

> maybe define an instance object like https://github.com/csi-addons/kubernetes-csi-addons/blob/main/internal/controller/replication.storage/volumereplication_controller.go#L616
and then write helper functions with it ?

I don't see any use case of defining such an instance object here, the only helper function here is the `getPVCFromDataSource` and it doesn't makes sense to add this extra type for a single function.

> The reconcile function is >250 lines.

It's a bit long I agree, but is there any rules defined for the lenght of the reconcile function?
Secondly, the operations that are happening in the reconcile function are very basic and can't be broken into individual helper functions. Those, that can be broken are already done. If you have any suggestions for it, I would be happy to know.

Reply:
Updated the code to move the creation of CRs into their individual functions.
Now it's < 250 :)